grpcproxy: support cancel watcher

xiang90 opened the pull request:
We do not wait for the cancellation from actual etcd server,
but generate it at the proxy side. The rule is to return the
latest rev that the watcher has seen. This should be good
enough for most use cases if not all.

/cc @heyitsanthony

heyitsanthony commented on proxy/grpcproxy/watcher_groups.go:
```diff
@@ -54,9 +55,10 @@ func (wgs *watchergroups) addWatcher(rid receiverID, w watcher) {
 }
 
-func (wgs *watchergroups) removeWatcher(rid receiverID) bool {
+func (wgs *watchergroups) removeWatcher(rid receiverID) (int64, bool) {
```
just return the revision since -1 is equivalent to false now?

xiang90 replied:
ok.

xiang90 replied:
after a second thought while changing the code, i feel we should keep the bool. Or we will have rev != -1 all over the place. and it is really clear why we need to compare rev to -1.

heyitsanthony replied:
ok fair enough

heyitsanthony commented on proxy/grpcproxy/watch.go:
```diff
@@ -218,16 +233,20 @@ func (sws *serverWatchStream) maybeCoalesceWatcher(ws watcherSingle) bool {
 }
 
-func (sws *serverWatchStream) removeWatcher(id int64) {
+func (sws *serverWatchStream) removeWatcher(id int64) (int64, bool) {
```
just return the revision since -1 <=> false?

heyitsanthony commented on proxy/grpcproxy/watcher_groups.go:
```diff
@@ -54,9 +55,10 @@ func (wgs *watchergroups) addWatcher(rid receiverID, w watcher) {
 	watchg.add(rid, w)
 	go watchg.run()
 	groups[w.wr] = watchg
+	wgs.idToGroup[rid] = watchg
```
should there be a `delete(wgs.idToGroup, rid)` somewhere in the remove path?

xiang90 commented:
@heyitsanthony I reworked this pr a little bit. I moved the cancelling logic into the remove func which is more consistent with the creation func. Also I removed the ctrlCh since we do not need it.

heyitsanthony commented:
lgtm
